runtimeVM: Fix shimv2 binary name construction

fidencio opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/master/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

/kind bug


#### What this PR does / why we need it:

shim v2 expected the runtime name to be in the following format:
```
    ($dir.)?$prefix.$name.$version
    -------- ------ ----- ---------
        |     |     |      |
        v     |     |      |
"/usr/local/bin"    |      |
  (optional)  |     |      |
              v     |      |
       "containerd.shim."  |
                    |      |
                    v      |
               "kata-qemu" |
                           v
                          "v2"
```

The current code simply replaces all the "-" by ".", which ends up
working well as long as the $name doesn't have a "-" in it, which is, by
luck, what's done by the Kata Operator.  However, whenever a name has
"-" in it, which we can see happening on upstream kata-deploy - which
uses $name as kata-qemu, for instance - it'll simply make containerd
library confused, not being able to invoke the binary.

With everything said above in mind, let's introduce a specialized
function to build the shim v2 path.

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

Fixes: #4589

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

fidencio commented on internal/oci/runtime_vm.go:
```diff
@@ -52,6 +53,33 @@ type containerInfo struct {
+	const binaryVersion = "v2"
+
+	runtimePrefix := filepath.Dir(path)
+	runtimeName := strings.TrimLeft(strings.TrimRight(filepath.Base(path), expectedVersion), expectedPrefix)
```
Oh, I thought Trim{Left,Right} would take prefixes instead of cutsets.  Okay, TIL and this is something I need to update.

fidencio commented:
/area vm

fidencio commented:
/cc @fgiudici

fidencio commented:
/retest

fidencio commented:
/test

fidencio commented:
/retest

fgiudici commented on internal/oci/runtime_vm.go:
```diff
@@ -52,6 +53,33 @@ type containerInfo struct {
+	runtimeDir := filepath.Dir(path)
+	runtimeName := strings.SplitAfter(strings.Split(filepath.Base(path), expectedVersion)[0], expectedPrefix)[1]
+
+	return filepath.Join(runtimeDir, fmt.Sprintf("%s.%s.%s", binaryPrefix, runtimeName, binaryVersion))
```
Here we enforce the containerd v2 shim prefix and suffix (also if not present in the passed path).

haircommander replied:
do we verify the file exists later?

fidencio replied:
We, as CRI-O, don't, but do we need?
Once this name is built it's passed to https://github.com/containerd/containerd/blob/08d765af954cd5355b7207fe628d06ed5a97c9d4/runtime/v2/shim/util.go#L41, which does the validation here https://github.com/containerd/containerd/blob/08d765af954cd5355b7207fe628d06ed5a97c9d4/runtime/v2/shim/util.go#L61-L94

fidencio replied:
Oh, actually CRI-O does check the existence of the binary during the validation step.

fgiudici reviewed (commented):
Here we assume that the VM runtime binary is in the form we expect (which is the containerd shimv2 format, i.e., containerd-shim-$RUNTIMENAME-v2). If that assumption is not met (wrong config file?) we can tweak the name or panic (see inline comments).
Maybe we already check the runtime VM name somewhere earlier... or maybe we don't need to deal with this here anyway (which would mean deal with unknown VM type runtimes or wrong configuration files): if so, just ignore the comments please.
Apart from this, nice patch, LGTM!

haircommander commented on internal/oci/runtime_vm.go:
```diff
@@ -52,6 +53,33 @@ type containerInfo struct {
 	cio *cio.ContainerIO
 }
 
+func buildShimV2Path(path string) string {
```
I think this may be better suited for config validation step in pkg/config/config.go

that way, we fail early if a user configured vm runtime incorrectly

fidencio replied:
@haircommander, let me see if I understand your suggestion correctly.  Your suggestion seems to be along the lines of what @fgiudici suggested here: https://github.com/cri-o/cri-o/pull/4590#discussion_r580187812, and I'm totally in for that.

With that in mind, we'd still need to build the "ShimV2Path", as what we pass to containerd is **not** exactly the same as the path of the binary.

For instance, for let's take the `containerd-shim-kata-v2` binary as example:
```
containerd-shim-kata-v2
                ----
                  |
                  v
         this should not change, and
    every other "-" should be replaces by "."
```
This will lead us to "containerd.shim.kata.v2" as the name passed to `containerd`.


Now, let's take `containerd-shim-kata-qemu-v2`, binary as example:
```
containerd-shim-kata-qemu-v2
                ---------
                    |
                    v
             this should not change, and
        every other "-" should be replaces by "."
```
This will lead us to "containerd.shim.kata-qemu.v2" as the name passed to `contained`.

So, AFAIU, we could validate that the binary name would be something like:
`(/\\w\/-+)+/containerd-shim-\\w+-v2` as part of the pkg/config/config.go (and, again, that does make sense), but we would still have to convert the `containerd-shim-$something-v2" to `containerd.shim.$something.v2`.

Based on the explanation above, do you still think the whole `buildShimV2Path` could be done at the config validation step in `pkg/config/config.go`?

haircommander commented:
I am fine with this here, and we can add validation in config later if we want

before we move forward with this PR though I'd like to see some unit tests :)

haircommander commented:
that button is too close to the comment button :upside_down_face:

fidencio commented:
@haircommander, @fgiudici,

Sorry, it took me longer than I wanted to be able to revisit this series.
I've addressed your comments, or I think I did, in this new series by:
* Enforcing a "containerd-shim-%s-v2" pattern (more or less), during the RuntimePath() validation, when using the "VM" runtime type;
* I've implemented some tests, the basic ones for the cases I'm trying to cover here;

Also as part of this series, I  ended up increasing the unit tests we already have so we cover other pieces related to the "VM" runtime type.

Ah, I must admit I'm not that comfortable with the `BuildContainerdBinaryName` being exposed as part of the `oci` package, but there's not much I can do, as that's the way to have it tested.  Meanwhile, let's count that the next developer using that piece of code will read the comments I added to that method.

fidencio commented:
And now the linter should be happier.

haircommander commented:
/approve

LGTM
/override ci/kata-jenkins

fgiudici commented:
LGTM

fidencio commented:
/retest

fidencio commented:
/test integration_cgroupv2

fidencio commented:
/retest

fidencio commented:
/retest

fidencio commented:
/retest

saschagrunert reviewed (approved):
/lgtm

fidencio commented:
/retest

TomSweeneyRedHat commented on internal/oci/oci.go:
```diff
@@ -477,3 +479,41 @@ type ExecSyncError struct {
 	return fmt.Sprintf("command error: %+v, stdout: %s, stderr: %s, exit code %d", e.Err, e.Stdout.Bytes(), e.Stderr.Bytes(), e.ExitCode)
 }
+
+// BuildContainrdBinaryName() is responsible for ensuring the binary passed will
```
nit of a nit
```suggestion
// BuildContainerdBinaryName() is responsible for ensuring the binary passed will
```

fidencio replied:
Fixed, thanks,

TomSweeneyRedHat commented:
LGTM
other than the nit which could be cleaned up later.

fidencio commented:
@saschagrunert, @haircommander, I may a new `/lgtm` label here. :-)

haircommander commented:
/lgtm

haircommander commented:
/test e2e-gcp

fidencio commented:
/retest

fidencio commented:
/test e2e-gcp

fidencio commented:
/retest

fidencio commented:
/test e2e-gcp
